src/decode.rs

```diff
@@ -0,0 +1,52 @@
+// Copyright 2019 King's College London.
+// Created by the Software Development Team <http://soft-dev.org/>.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use crate::{Expect, Index, Mir, FORMAT_VERSION};
+use rmp_serde::{
+    decode::{self, ReadReader},
+    Deserializer,
+};
+use serde::Deserialize;
+use std::io::Read;
+
+/// A decoder.
+pub struct Decoder<'a> {
+    deser: Deserializer<ReadReader<&'a mut dyn Read>>,
+    index: Index,
+}
+
+impl<'a> Decoder<'a> {
+    /// Returns a new decoder which will deserialise from `read_from` and also an `Index` informing
+    /// the consumer of the expected contents.
+    pub fn new(read_from: &'a mut dyn Read) -> Result<(Index, Self), decode::Error> {
```

Comment:
This feels like it's more of a `from_` method than a `new` method, because I can vaguely imagine other things than a file that we want to deserialise (e.g. a string).

Reply:
Agreed.